### DISCUSSION TITLE
Mirai 文档建议

### DISCUSSION THREAD

**Discussion by Him188** [mamoe/mirai#848] [Ideas]
如果你对现有文档有任何建议, 可以在这里提出.

**Comment by litlum** [comment]
把Contact和Bot的类图画一下吧

**Reply by Him188** [reply]
https://github.com/mamoe/mirai/blob/dev/docs/Contacts.md 如何

**Comment by KADGang** [comment]
可以试着征收其他开发者的一些 保姆级教学吗，感觉相比于直接看文档和源码对于某些新手来说还是有些晦涩，如果能有一些十分简单的教学去先让新手们做一点简单的东西在慢慢往文档和源码上面指引会更加好吧，或者试着做mirai的wiki啥的

**Reply by Him188** [reply]
可以，欢迎提交

**Reply by KADGang** [reply]
好的 如果将来有余力会尽力做贡献的

**Comment by liaosunny123** [comment]
Mirai社区还接受其他启动器吗?我个人维护一个可视化的Mirai辅助启动器，方便用户操作和运行，目前仍在维护。

**Reply by Him188** [reply]
接受，在稳定后欢迎回复

**Reply by liaosunny123** [reply]
> 接受，在稳定后欢迎回复

好的，谢谢了，目前已经完成对mirai的一键启动和http api的适配和httpapi错误自动处理，目前在做插件中心，已经可以正常使用了，这是项目地址:https://www.github.com/liaosunny123/epicmomiraig
不过一直只有十几个人用，没办法让其他人知道这个项目

**Reply by Him188** [reply]
欢迎在[论坛](https://mirai.mamoe.net)发布你的开发进度

**Comment by chrysoljq** [comment]
可以展示各个基本功能的用法或者说api吗

**Reply by StageGuard** [reply]
开发文档介绍基本功能很详细了，细节功能可以借助 IDEA 补全来看。

**Reply by Him188** [reply]
Message 文档正在更新中 #1088 

会不断根据最近收到的 issue 和 GitHub 开源的插件的开发情况来调整文档在简洁与完善的平衡点

**Comment by Nambers** [comment]
希望加个xml构造器的大概用法什么的
如<https://github.com/khjxiaogu/MiraiSongPlugin>这个项目里的音乐卡片或其他
目前应该是只能自己发送卡片然后机器人解析在文本发在日志从而获取到xml里面的参数和格式什么的，希望mirai能给个大概的xml的一些介绍?毕竟就自己分析比较费时间，而mirai这边可能有经验吧hhh
或者之后我有时间可能会做个,如果mirai这边没有的话，那我能pr进doc吗hhh

**Reply by Him188** [reply]
mirai 没有分析过 xml，只提供发送接受功能

**Reply by Nambers** [reply]
> mirai 没有分析过 xml，只提供发送接受功能

okok 之前好像看见过有，可能看错了

**Reply by Him188** [reply]
那个是'随便'写的一个构建器, 只适用于很小的情况. 要使用还是要自己写

**Reply by Nambers** [reply]
> 那个是'随便'写的一个构建器, 只适用于很小的情况. 要使用还是要自己写

okok hhh，据说json更稳定点，我就去构造json发送了https://github.com/Nambers/MiraiCP/blob/master/doc/example.md#%E5%8F%91%E9%80%81lightapp%E4%BB%A5%E5%B0%8F%E7%A8%8B%E5%BA%8F%E5%BD%A2%E5%BC%8F%E5%8F%91%E9%80%81%E5%8D%A1%E7%89%87

**Comment by Nambers** [comment]


**Reply by PeratX** [reply]
这是协议原文，不可修改。

**Reply by Nambers** [reply]
ok

**Comment by xuelianbobo** [comment]
想请教如何处理收到的文件类消息，多谢

**Reply by Him188** [reply]
> 想请教如何处理收到的文件类消息，多谢

目前还没有支持，见 #805

**Comment by Agoines** [comment]
想请教如何获取他人发送的消息

**Reply by Agoines** [reply]
因为原有的 QQ 和 TIM 体型太大，所以尝试通过mirai作为内核，在Android端创建一个能简单处理并回复他人消息的客户端。通过阅读文档后，仍未理解如何获取他人发送的消息，所以才有了上述的求助，望解答

**Reply by Nambers** [reply]
使用mirai core构建mirai服务？就登陆个bot然后监听事件吧 从bot.eventchannrl

**Reply by Agoines** [reply]
> 使用mirai core构建mirai服务？就登陆个bot然后监听事件吧 从bot.eventchannrl

感谢解答，我去试试

**Reply by Nambers** [reply]
> > 使用mirai core构建mirai服务？就登陆个bot然后监听事件吧 从bot.eventchannrl
> 
> 感谢解答，我去试试

我记得mirai有个示例项目mirai hello world，不知道你想要的是不是这种，就用mirai core构建个软件，不依赖于mirai console

**Reply by Agoines** [reply]
我在构建一个类似的android端项目，hello world 是一个新建一个 Bot，但是似乎没有其他详细的文档。因为第一次尝试去制作一个完善的Android端应用，也是第一次使用mirai这个项目，也是第一次尝试使用 github，所以在查阅 mirai 文档，也有很多不太明白，我希望这方面的，能多一些保姆级教程

**Reply by Nambers** [reply]
> 我在构建一个类似的android端项目，hello world 是一个新建一个 Bot，但是似乎没有其他详细的文档。因为第一次尝试去制作一个完善的Android端应用，也是第一次使用mirai这个项目，也是第一次尝试使用 github，所以在查阅 mirai 文档，也有很多不太明白，我希望这方面的，能多一些保姆级教程

先构建一个bot，然后在下面用bot.eventchannel.subscribealawy<>监听事件，如群消息事件或私聊信息事件

**Reply by yzqd9527** [reply]
你有没有了解过MiraiAndroid这个项目呢

**Reply by Nambers** [reply]
> 你有没有了解过MiraiAndroid这个项目呢

他貌似是想做个仿qq客户端，而那个项目好像是在android上跑plugin？（我没怎么看那个项目

**Reply by Agoines** [reply]
> 你有没有了解过MiraiAndroid这个项目呢

没怎么了解，我是打算，仿成客户端，那个似乎是加载插件

**Comment by NekoWs** [comment]
部分消息函数没有列出，我废了好大劲才在注释里找到撤回消息
还有MessageChain怎么转换为MessageSource

**Reply by Nambers** [reply]
.recall()，然后转换到MessageSource我记得注释里有，然后如果没记错是MessageChain['Source']
如果没记错是在MessageChain定义里的注释

**Reply by NekoWs** [reply]
是MessageChain直接.recall()吗？我用Java，没有这个函数啊

**Reply by Nambers** [reply]
> 是MessageChain直接.recall()吗？我用Java，没有这个函数啊

喔在kt里如果没记错是messagechain.recall()，java可能不是，java没有kt的扩展函数

**Reply by NekoWs** [reply]
我现在用的是Mirai.getInstance().recallMessage(bot , MessageSource);

**Reply by NekoWs** [reply]
> > 是MessageChain直接.recall()吗？我用Java，没有这个函数啊
> 
> 喔在kt里如果没记错是messagechain.recall()，java可能不是，java没有kt的扩展函数

那Java用什么方法，只能用Mirai.getInstance().recallMessage(Bot , MessageSource);吗，那MessageChain转换到MessageSource呢？

**Reply by Nambers** [reply]
> > > 是MessageChain直接.recall()吗？我用Java，没有这个函数啊
> > 
> > 
> > 喔在kt里如果没记错是messagechain.recall()，java可能不是，java没有kt的扩展函数
> 
> 那Java用什么方法，只能用Mirai.getInstance().recallMessage(Bot , MessageSource);吗，那MessageChain转换到MessageSource呢？

如果没记错是MessageChain['Source'] 上面信息有写

**Reply by Him188** [reply]
消息相关将会在 2.6 补充各种操作方法和处理思想 #1088

**Reply by Him188** [reply]
如果可能, 还会在 IntelliJ 为 Java 添加一些 LIve Template 来提供联想

**Comment by Widecss** [comment]
作为刚接触的用户，我希望能在消息发送中加上外部文件发送的例子或指引（ 图片等 ），
我在寻找发送图片的方式，通过注释知道了发送方法可以接受 Message 参数，而 Image 是其子类，
看了半天 Image 类，在其中只能找到由 ID 构造的方法，有些让人费解，而文档也未说明应该如何发送图片，
最终才在 Contact 类翻到了 uploadImage 方法。
虽然我的思路不太对，但是不一定只有我走进了这种不正确的思路，所以希望可以修改一下文档。

**Reply by Him188** [reply]
将会作为 #1088 的任务之一

**Reply by Widecss** [reply]
> 将会作为 #1088 的任务之一

好的感谢

**Comment by Agoines** [comment]
在测试学习中，Android Studio 对mirai-android的混淆，导致了闪退，在查阅 mirai 的 issues 中没有这个问题答案。猜测可能 mirai-android 混淆导致的。

希望可以提供 ProGuard 的混淆方案，或者单独将 mirai-android 分为一个仓库，对 mirai-android 相关的问题更好分类。

**Reply by Him188** [reply]
#992 
现在我们缺乏 ProGuard 配置经验而搁置了这个任务, 欢迎 PR

Mirai Android 在独立项目, <https://github.com/mzdluo123/miraiandroid>

**Reply by Agoines** [reply]
> #992 
> 现在我们缺乏 ProGuard 配置经验而搁置了这个任务, 欢迎 PR
> 
> Mirai Android 在独立项目, <https://github.com/mzdluo123/miraiandroid>

上文的 mirai-android 指的是 mirai-core-android 这样的mirai-core 分支 https://search.maven.org/search?q=a:mirai-core-android. 因为名字相似，因此带来歧义，表示抱歉。

我在尝试制作可视化的 mirai 机器人，并且在练习配置 ProGuard. 如果，有成果就会提交 PR.

**Reply by Him188** [reply]
`mirai-core-android` 是 Kotlin MPP 的一个 target, 无法分割

**Reply by Him188** [reply]
如果只是对 Android 平台有问题, 未来加一个 issue tag 就可以了, 分仓库会造成很大的开发和管理的麻烦, 当 issue 比较多的时候才是'划算'的

**Reply by Agoines** [reply]
> 如果只是对 Android 平台有问题, 未来加一个 issue tag 就可以了, 分仓库会造成很大的开发和管理的麻烦, 当 issue 比较多的时候才是'划算'的

明白了，我去研究下 ProGuard, 争取能为开源提供自己一份力

**Comment by TihPct** [comment]
我有一个问题想请教一下，MemberJoinEvent.Invite方法并不能获取到邀请者的信息，只能获取到被邀请者的信息，可有其他API可获取到吗
谢谢！

**Reply by Him188** [reply]
![image](https://user-images.githubusercontent.com/12100985/112607822-94379600-8e54-11eb-9e31-1ff8a12547d9.png)

它是有 `invitor` 的

**Reply by TihPct** [reply]
![image](https://user-images.githubusercontent.com/51705716/112614100-d3b5b080-8e5b-11eb-847b-f943047fc092.png)
我想应该是我的版本太低了
我使用的是1.2.3的mirai-core-qqandroid-1.2.3

**Reply by TihPct** [reply]
![image](https://user-images.githubusercontent.com/51705716/112614502-4de63500-8e5c-11eb-9cca-3694deb2eb66.png)

**Comment by eiriksgata** [comment]
想把当前GithubPage项目文档改为使用docsify，这样具有更好的导航以及视图浏览，全局搜索关键字也会变得更为容易。

**Reply by eiriksgata** [reply]
这边稍微改了一下，这是预览效果：https://eiriksgata.github.io/mirai/#/

**Comment by sandtechnology** [comment]


**Reply by jerrita** [reply]
嗯嗯，刚刚发现了，谢谢^_^

**Comment by sandtechnology** [comment]
https://github.com/mamoe/mirai/issues/1173#issuecomment-817766600

>经过阅读`BotConfiguration.kt`已基本了解网络日志，应该是包含“Net”前缀的日志。但其他对于本项目不熟悉的人很可能会一头雾水，建议在issue模板中添加对于网络日志特征的描述，这样给双方都能带来方便。

**Comment by imotosaiko** [comment]
想请教一下文档给出的事件列表中的OtherClientMessagePreSendEvent和OtherClientMessagePostSendEvent是已经弃用的事件类吗？为什么在开发的时候idea无法找到这两个事件类呢？
![图片](https://user-images.githubusercontent.com/71994020/129407480-0d0bb8e5-afda-4eb3-9781-aed53de03953.png)
![图片](https://user-images.githubusercontent.com/71994020/129408687-7bc2d48b-c756-4b80-9847-8e7d29131826.png)

**Reply by imotosaiko** [reply]
已解决，可以通过FriendMessageSyncEvent实现功能（文档里完全没写啊喂！！（额，要是我眼瞎了那就另说（

**Comment by eric2788** [comment]
请问哪里可以看javadoc文件？

**Comment by hundun000** [comment]
当插件开发者为具体插件编写用户手册时，需要描述智能参数解析对应的参数，可以有哪些表达格式。例如：

> 参数1    类型: User    说明: 用mirai规定格式的UserID形式表达，详见下文
> 
> - `botId.groupId.memberId`
> - `botId.groupId.memberCard` (模糊搜索, 寻找最优匹配)
> - `~` (指代指令调用人自己. 仅聊天环境下)
> - `botId.groupId.$` (随机成员. )
> - `botId.friendId`
> 当处于一个群内时, `botId` 和 `groupId` 参数都可省略
>        当只登录了一个 [Bot] 时, `botId` 参数可省略

虽然插件开发者可以从CommandArgumentParserBuiltins.kt的注释里复制描述贴到插件用户手册里，但是可能错漏，也比较繁琐。如果官方用户手册补充相关说明，插件用户手册链接引用，应该更合适。

**Comment by LinusOrange** [comment]
亲测安装路径不能有中文，建议添加到用户手册里？

**Comment by Riaruuu** [comment]
我有这么几个问题，
第一个是我是用springBoot做的一个B/S的程序， 在浏览器上对机器人做一些基本的设置，然后登录的时候不会有那个验证码弹窗出来，我每次都是单独写了个登录的程序登录一遍信任了设备才能再运行服务端；我在想是不是可以加个扫码登录的功能，暴露两个接口，一个是提供二维码code的，我可以根据code再页面展示，另一个相当于回调，我扫码点击登录再调用这个接口，当然我不清楚QQ的api不知道是否可行。

第二个是我每次发图片， 如果是本地的图片还好，如果是网络图片url，我还需要先下载下来，再调用upload方法拿到id在发送， 能不能直接根据url发送啊

**Reply by Him188** [reply]
扫码登录 #1281 

URL发图片，你可以自己写一个工具方法，因为core不知道你使用哪个 HTTP 库之类的。
